### PR TITLE
nixos/gitlab-runner: leave runner names as is

### PR DESCRIPTION
**Pull request by misuzu**
###### Description of changes
Currently a runner name looks like this: `$name_$hostname_$somehash`. Users often complaining that they want stable name instead.
With this change a runner name is just `$name` and `$somehash` part is stored separately in a file. This behaviour is difficult to achieve properly using bash, so I've added a Python script to figure out what runners should be registered and unregistered.
I've tried adding `x-nixos-name` for each runner directly in `config.toml`, but `gitlab-runner` removes unknown keys 🤡
Feel free to offer a better approach 😄 
We can also stick to the status quo.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by teto** [comment]
I managed to have stable names using --description thanks to https://github.com/NixOS/nixpkgs/pull/170396 but this seems to have broken the unregistering step. Just as a datapoint. Will have a look later

File: nixos/modules/services/continuous-integration/gitlab-runner.nix [hunk 1]
```diff
@@ -35,12 +99,20 @@ let
...
       # current and desired state
-      NEEDED_SERVICES=$(echo ${concatStringsSep " " (attrNames hashedServices)} | tr " " "\n")
+      CURRENT_SERVICES=$(cat $REGISTERED_STATE_FILE || true)
+      DESIRED_SERVICES=$(echo ${escapeShellArg (concatStringsSep "\n" hashedServices)})
       REGISTERED_SERVICES=$(gitlab-runner list 2>&1 | grep 'Executor' | awk '{ print $1 }')
 
       # difference between current and desired state
-      NEW_SERVICES=$(grep -vxF -f <(echo "$REGISTERED_SERVICES") <(echo "$NEEDED_SERVICES") || true)
-      OLD_SERVICES=$(grep -vxF -f <(echo "$NEEDED_SERVICES") <(echo "$REGISTERED_SERVICES") || true)
+      NEW_SERVICES=$(${getRunnerServices} register --current "$CURRENT_SERVICES" --desired "$DESIRED_SERVICES" --registered "$REGISTERED_SERVICES")
```
Comment (teto) [inline nixos/modules/services/continuous-integration/gitlab-runner.nix:107]: I find the semantic awkward, it's the same call to register than for unregister and we still have a shell script that calls gitlab-runner unregister, what's the point of calling `${getRunnerServices} unregister` then ?

File: nixos/modules/services/continuous-integration/gitlab-runner.nix [hunk 2]
```diff
@@ -4,14 +4,76 @@ with lib;
...
-      service)
-    cfg.services;
+  hashedServices = mapAttrsToList
+    (name: service: "${name}:${hashString "md5" (unsafeDiscardStringContext (toJSON service))}")
+     cfg.services;
   configPath = "$HOME/.gitlab-runner/config.toml";
+  getRunnerServices = (pkgs.writers.writePython3 "get-runner-services" {
+    libraries = with pkgs.python3Packages; [ click ];
+  } ''
+    import click
```
Comment (teto) [inline nixos/modules/services/continuous-integration/gitlab-runner.nix:14]: we can achieve a similar friendliness with just argparse. Better not to depend on external package.

File: nixos/modules/services/continuous-integration/gitlab-runner.nix (hunk 1 above)
Comment (misuzu) [reply nixos/modules/services/continuous-integration/gitlab-runner.nix:107]: `${getRunnerServices} unregister` outputs a list of services which needs to be unregistered.

**Comment by misuzu** [comment]
I really don't like this approach, the best option would be to fix this [upstream](https://gitlab.com/gitlab-org/gitlab-runner/-/issues/5024).

**Comment by teto** [comment]
so I tried to dabble with a simialr issue: when you register a runner with spaces in its name (possible if you use description submodule) then you runners never get unregistered because our bash scripts assume no space in names.
I've spent way too much time trying to fix this in bash but between the funky diff compareasons, gitlab-runner list displaying random invisible characters that break my regex etc I would like to put some python as well.
My PR idea is to:
- always encode the name, even in "description" (currently not the case, which breaks the whole logic of the module, my bad).
- rely on gitlab-runner --log-format json list to get the registered list
- unregister them via their tokens. I dont know why but my module fails unregistering runners even though the name is correct, it works when I launch it manually etc. Plus using token is more precise when looking at the logs. 

Some of my suffering logged at https://gitlab.com/gitlab-org/gitlab-runner/-/issues/29350

**Pull request closed without merging**